File: Homework #1/Solutions/cpp/quicksort.cpp
/*
 Matthew Romano
 EECS 592
 Problem #2 - quicksort.cpp
*/

#include <iostream>
#include <fstream>      // file input/output
#include <string>       // string
#include <vector>       // vector


// Class Declarations
class entry {
 public:
  entry(void) {}
  entry(std::string lN, std::string fN, int g) : lastName(lN),
                                                 firstName(fN),
                                                 grade(g) {}

  std::string lastName;
  std::string firstName;
  int grade;
};

// Function Declarations
void quicksort(std::vector<entry>* entries, int lo, int hi);
int partition(std::vector<entry>* entries, int lo, int hi);

int main() {
    // File Operation Variables
    std::ifstream inFile("unsorted.txt");
    std::ofstream outFile("sorted.txt");
    std::string line, last, first, grade;
    char delim = 44;  // comma
    size_t p, q;

    // Entries to be sorted by quicksort() function
    std::vector<entry> entries;

    // 1) Read in the data
    if (inFile.is_open()) {
        while (getline(inFile, line)) {
            p = line.find_first_of(delim);
            q = line.find_last_of(delim);

            last = line.substr(0, p);
            first = line.substr(p+2, q-(p+2));
            grade = line.substr(q+2);

            entry anEntry(last, first, stoi(grade));
            entries.push_back(anEntry);
        }
        inFile.close();
    } else {
        std::cout << "Unable to open file";
    }

    // 2) Sort the data
    quicksort(&entries, 0, entries.size()-1);

    // 3) Write the sorted data to the output file
    for (std::vector<entry>::iterator it = entries.begin();
                                      it != entries.end(); ++it) {
        outFile << (*it).lastName  << ", " <<
                   (*it).firstName << ", " <<
                   (*it).grade     << std::endl;
                       }
    outFile.close();

    return 0;
}



void quicksort(std::vector<entry>* entries, int lo, int hi) {
    if (lo < hi) {
        int p = partition(entries, lo, hi);
        quicksort(entries, lo, p-1);
        quicksort(entries, p+1, hi);
    }
}

int partition(std::vector<entry>* entries, int lo, int hi) {
    entry temp;

    int pivot = (*entries)[hi].grade;
    int i = lo -1;

    for (int j = lo; j < hi; j++) {
        if ((*entries)[j].grade < pivot) {
            i++;
            temp = (*entries)[i];
            (*entries)[i] = (*entries)[j];
            (*entries)[j] = temp;
        }
    }
    if ((*entries)[hi].grade < (*entries)[i+1].grade) {
        temp = (*entries)[i+1];
        (*entries)[i+1] = (*entries)[hi];
        (*entries)[hi] = temp;
    }

    return i+1;
}

File: Project0/Solutions/python/quicksort.py
""" quicksort - Problem #2

    Matthew Romano
    eecs592
"""

import csv


class Entry:
    """ Class to hold details for each entry """

    def __init__(self, lastname, firstname, grade):
        self.lastname = lastname
        self.firstname = firstname
        self.grade = grade
    def __lt__(self, other): # overload < operator
        return self.grade < other.grade


def partition(entries, low, high):
    """ Partitions entries using the Lomuto partition scheme.

        https://en.wikipedia.org/wiki/Quicksort#Lomuto_partition_scheme
     """

    pivot = entries[high]
    i = low-1
    for j in range(low, high):
        if entries[j] < pivot:
            i += 1
            entries[i], entries[j] = entries[j], entries[i]
    if entries[high] < entries[i+1]:
        entries[high], entries[i+1] = entries[i+1], entries[high]
    return i+1


def quicksort(entries, low, high):
    """ Recursive quicksort function

        Sorts entries from least to greatest between the low and high indices
     """
    if low < high:
        p_index = partition(entries, low, high)
        quicksort(entries, low, p_index-1)
        quicksort(entries, p_index+1, high)


def main():
    """ Main function

        1) Reads in the input file
        2) Calls the quicksort algorithm
        3) Writes the sorted array to an output file
    """

    entries = []

    # 1) Read in the data
    with open('unsorted.txt', 'r') as infile:
        csvreader = csv.reader(infile)
        for line in csvreader:
            lastname = line[0]
            firstname = line[1][1:]
            grade = int(line[2][1:])
            an_entry = Entry(lastname, firstname, grade)
            entries.append(an_entry)

    # 2) Sort the data
    quicksort(entries, 0, len(entries)-1)

    # 3) Write the sorted data to the output file
    with open('sorted.txt', 'w') as outfile:
        for entry in entries:
            outfile.write(entry.lastname + ', ' + entry.firstname + ', ' + str(entry.grade) + '\n')


if __name__ == "__main__":
    main()
